src/runtime.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "common.h"
+#include "runtime.h"
+
+static git_runtime_shutdown_fn shutdown_callback[32];
```

Comment:
It's a wee bit unfortunate we now have a static limit on shutdowns again. But given that overall this PR is quite an improvement, I'm fine with this tradeoff.